NWNXLib/Services/Commands/Commands.cpp

```diff
@@ -0,0 +1,144 @@
+#include "Services/Commands/Commands.hpp"
+#include <stdio.h>
+#include <dlfcn.h>
+
+namespace NWNXLib {
+namespace Services {
+
+static Commands *g_commands;
+
+Commands::Commands()
+{
+    g_commands = this;
+    RegisterCommand("nwnx_commands_test", [](std::string& s){ LOG_NOTICE("Selftest command. Args: '%s'", s.c_str()); });
+}
+
+Commands::~Commands()
+{
+    g_commands = nullptr;
+
+}
+
+bool Commands::RegisterCommand(const std::string& cmd, CommandFunc func)
+{
+    auto it = m_commandMap.find(cmd);
+    if (it == m_commandMap.end())
+    {
+        m_commandMap[cmd] = func;
+        LOG_INFO("Registering command '%s'", cmd.c_str());
+        return true;
+    }
+    else
+    {
+        LOG_WARNING("Command '%s' already registered", cmd.c_str());
+        return false;
+    }
+}
+
+void Commands::UnregisterCommand(const std::string& cmd)
+{
+    LOG_INFO("Unregistering command '%s'", cmd.c_str());
+    m_commandMap.erase(cmd);
+}
+
+// This runs on a side thread that processes input.
+bool Commands::ScheduleCommand(std::string cmdline)
+{
+    auto trim = [](std::string &s)
+    {
+        s.erase(0, s.find_first_not_of(" \n\r\t"));
+        s.erase(s.find_last_not_of(" \n\r\t") + 1);
+    };
+
+    trim(cmdline);
+    std::string cmd, args;
+    size_t space = cmdline.find(' ');
+
+    cmd = cmdline.substr(0, space);
+    args = (space == std::string::npos) ? "" : cmdline.substr(space, std::string::npos);
+
+    trim(cmd); trim(args);
+
+    auto it = m_commandMap.find(cmd);
+    if (it != m_commandMap.end())
+    {
+        std::lock_guard<std::mutex> guard(m_queueLock);
```

Comment:
This mutex needs to cover the `m_commandMap.find(cmd)`, else there's a race condition where you may try to search the vector while the other thread is in the process of clearing it.

Reply:
Good spot, resolved.